src/circuit/server/index.ts

```diff
@@ -0,0 +1,427 @@
+    }
+  }
+
+  async makeReservation (
+    remotePeer: PeerId,
+    expire: bigint
+  ): Promise<Reservation> {
+    const addrs = []
+
+    for (const relayAddr of this.addressManager.getAddresses()) {
+      addrs.push(relayAddr.bytes)
+    }
+
+    const voucher = await RecordEnvelope.seal(new ReservationVoucherRecord({
+      peer: remotePeer,
+      relay: this.peerId,
+      expiration: Number(expire)
+    }), this.peerId)
+
+    return {
+      addrs,
+      expire,
+      voucher: voucher.marshal()
+    }
+  }
+
+  async handleConnect ({ stream, request, connection }: HopProtocolOptions): Promise<void> {
+    const hopstr = stream.pb(HopMessage)
+
+    if (isRelayAddr(connection.remoteAddr)) {
+      log.error('relay reservation over circuit connection denied for peer: %p', connection.remotePeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+      return
+    }
+
+    log('hop connect request from %s', connection.remotePeer)
+
+    let dstPeer: PeerId
+
+    try {
+      if (request.peer == null) {
+        log.error('no peer info in hop connect request')
+        throw new Error('no peer info in request')
+      }
+
+      request.peer.addrs.forEach(multiaddr)
+      dstPeer = peerIdFromBytes(request.peer.id)
+    } catch (err) {
+      log.error('invalid hop connect request via peer %p %s', connection.remotePeer, err)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.MALFORMED_MESSAGE })
+      return
+    }
+
+    if (isRelayAddr(connection.remoteAddr)) {
+      log.error('hop connect request over circuit connection denied for peer: %p', connection.remotePeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+      return
+    }
+
+    if (!await this.reservationStore.hasReservation(dstPeer)) {
+      log.error('hop connect denied for destination peer %p not having a reservation for %p with status %s', dstPeer, connection.remotePeer, Status.NO_RESERVATION)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.NO_RESERVATION })
+      return
+    }
+
+    if ((await this.connectionGater.denyOutboundRelayedConnection?.(connection.remotePeer, dstPeer)) === true) {
+      log.error('hop connect for %p to %p denied by connection gater', connection.remotePeer, dstPeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+      return
+    }
+
+    const connections = this.connectionManager.getConnections(dstPeer)
+
+    if (connections.length === 0) {
+      log('hop connect denied for destination peer %p not having a connection for %p as there is no destination connection', dstPeer, connection.remotePeer)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: Status.NO_RESERVATION })
+      return
+    }
+
+    const destinationConnection = connections[0]
+    log('hop connect request from %s to %s is valid', connection.remotePeer, dstPeer)
```

Comment:
Is there a debug log level so that we don't log this all the time under normal operation?

Reply:
FWIW under normal operation these log lines are all no-ops, stuff is only logged if the `DEBUG` env var is set.